substrate/pkg/controller/substrate/cluster/addons/dns.go

```diff
@@ -0,0 +1,47 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package addons
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/awslabs/kit/substrate/pkg/apis/v1alpha1"
+	"github.com/awslabs/kit/substrate/pkg/controller/substrate/cluster"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type DNS struct {
```

Comment:
I'm a little curious if its worth colocating all of these addons into the same controller `addons.go`, since it's a bunch of interdependent calls to kubeadm. It may be marginally faster this way, but it appears as if there is more complexity than there is.

Reply:
Yeah we can explore but not all addons are straightforward as this on. The new functionality to deploy helm chart as addons I just added are little more complex.